server/src/server.jl

```diff
@@ -0,0 +1,164 @@
+    "get_value" => GetValue,
+    "get_value_ptr" => GetValuePtr,
+    "get_value_at_indices" => GetValueAtIndices,
+    "set_value" => SetValue,
+    "set_value_at_indices" => SetValueAtIndices,
+    "get_grid_type" => GetGridType,
+    "get_var_grid" => GetVarGrid,
+    "get_grid_rank" => GetGridRank,
+    "get_grid_size" => GetGridSize,
+    "get_grid_x" => GetGridX,
+    "get_grid_y" => GetGridY,
+    "get_grid_node_count" => GetGridNodeCount,
+    "get_grid_edge_count" => GetGridEdgeCount,
+    "get_grid_edge_nodes" => GetGridEdgeNodes,
+    "finalize" => Finalize,
+    "load_state" => LoadState,
+    "save_state" => SaveState,
+)
+
+mutable struct ModelHandler
+    model::Union{Wflow.Model,Nothing}
+end
+
+"Shutdown ZMQ server"
+function shutdown(s::Socket, ctx::Context)
+    @info "Shutting down Wflow ZMQ server on request..."
+    ZMQ.close(s)
+    ZMQ.close(ctx)
+end
+
+"Error response ZMQ server"
+function response(err::AbstractString, s::Socket)
+    @info "Send error response"
+    resp = Dict{String,String}("status" => "ERROR", "error" => err)
+    ZMQ.send(s, JSON3.write(resp))
+end
+
+"Status response ZMQ server"
+function response(s::Socket)
+    @info "Send status response"
+    resp = Dict{String,String}("status" => "OK")
+    ZMQ.send(s, JSON3.write(resp))
+end
+
+"Validate JSON request against mapped Struct"
+function valid_request(json)
+    for f in fieldnames(map_structs[json.fn])
+        if f ∉ keys(json)
+            return f
+            break
+        end
+    end
+end
+
+"""
+    wflow_bmi(s::Socket, handler::ModelHandler, f)
+
+Run a Wflow function through `wflow.bmi(f, handler.model)` and update Wflow Model `handler`
+if required, depending on return type of `wflow.bmi(f, handler.model)`.
+"""
+function wflow_bmi(s::Socket, handler::ModelHandler, f)
+    try
+        ret = wflow_bmi(f, handler.model)
+        if typeof(ret) <: Wflow.Model # update of Wflow model
+            handler.model = ret
+            response(s)
+        elseif isnothing(ret) # for SetValue and SetValueAtIndices
+            response(s)
+        else
+            @info "Send response including output from Wflow function `$(f.fn)`"
+            ZMQ.send(s, JSON3.write(ret))
+        end
+    catch e
+        @error "Wflow function `$(f.fn)` failed" exception = (e, catch_backtrace())
+        err = string(
+            "Wflow function `$(f.fn)` failed\n exception =\n ",
+            sprint(showerror, e, catch_backtrace()),
+        )
+        response(err, s)
+    end
+end
+
+"Start Wflow ZMQ Server"
+function start()
+    @info "Start Wflow ZMQ Server..."
+    
+    # initialize Wflow model handler
+    handler = ModelHandler(nothing)
+
+    # set up a ZMQ context, with optional port number (default = 5555) argument
+    context = Context()
+    socket = Socket(context, REP)
+    n = length(ARGS)
+    if n == 0
+        port = 5555
+    elseif n == 1
+        port = parse(Int, ARGS[1])
+    else
+        throw(
```

Comment:
Will `port` always stay the only argument? Otherwise requiring a flag like `--port 5555` would be more extensible.